src/main/java/frc/robot/subsystems/elevator/ElevatorIOReal.java

```diff
@@ -1,15 +1,96 @@
 
+import static edu.wpi.first.units.Units.Degrees;
+import static edu.wpi.first.units.Units.Meters;
+
+import com.ctre.phoenix6.controls.DutyCycleOut;
+import com.ctre.phoenix6.controls.MotionMagicVoltage;
+import com.ctre.phoenix6.controls.StrictFollower;
+import com.ctre.phoenix6.hardware.TalonFX;
 import edu.wpi.first.units.Distance;
 import edu.wpi.first.units.MutableMeasure;
+import edu.wpi.first.units.Units;
+import edu.wpi.first.wpilibj.DigitalInput;
+import edu.wpi.first.wpilibj.Servo;
+import frc.robot.Ports;
 
 public class ElevatorIOReal implements ElevatorIO {
+    private final Servo servo;
+    private final TalonFX mainMotor;
+    private final TalonFX auxMotor;
+
+    private final DigitalInput bottomSensor;
+    private final DigitalInput topSensor;
+
+    private final MotionMagicVoltage positionControl = new MotionMagicVoltage(0);
+    private final DutyCycleOut powerControl = new DutyCycleOut(0);
+
+    public ElevatorIOReal() {
+        mainMotor = new TalonFX(Ports.Elevator.MAIN_MOTOR);
+        auxMotor = new TalonFX(Ports.Elevator.AUX_MOTOR);
+        servo = new Servo(Ports.Elevator.SERVO);
+        bottomSensor = new DigitalInput(Ports.Elevator.BOTTOM_LIMIT_SWITCH);
+        topSensor = new DigitalInput(Ports.Elevator.TOP_LIMIT_SWITCH);
+
+        mainMotor.getConfigurator().apply(ElevatorConstants.MOTOR_CONFIGURATION);
+        auxMotor.getConfigurator().apply(ElevatorConstants.MOTOR_CONFIGURATION);
+
+        auxMotor.setControl(new StrictFollower(mainMotor.getDeviceID()));
+    }
+
     @Override
-    public void setPower(double power) {}
+    public void setPower(double power) {
+        mainMotor.setControl(powerControl.withOutput(power));
+    }
 
     @Override
-    public void setHeight(MutableMeasure<Distance> height) {}
+    public void setHeight(MutableMeasure<Distance> height) {
+        inputs.heightSetpoint.mut_replace(height);
+        mainMotor.setControl(positionControl.withPosition(height.in(Units.Meters)));
+    }
 
     @Override
-    public void updateInputs(ElevatorInputs inputs) {}
+    public void updateInputs(ElevatorInputs inputs) {
+        inputs.isBottom = atBottom();
+        inputs.isTop = atTop();
+        inputs.carriageHeight = Meters.of(mainMotor.getPosition().getValue()).mutableCopy();
+
+        inputs.gripperHeight.mut_replace(
+                inputs.carriageHeight.gt(ElevatorConstants.GRIPPER_HEIGHT)
+                        ? inputs.carriageHeight
+                                .mutableCopy()
+                                .mut_minus(ElevatorConstants.GRIPPER_HEIGHT)
+                        : Meters.zero());
+
+        inputs.servoAngle = Units.Degrees.of(servo.getAngle()).mutableCopy();
+    }
+
+    public boolean atTop() {
+        return topSensor.get();
+    }
+
+    public boolean atBottom() {
+        return bottomSensor.get();
+    }
+
+    public void flipServo() {
+        if (servo.get() == ElevatorConstants.SERVO_OPEN.in(Degrees)) {
```

Comment:
1. You'd probably not get the exact measurement, so `==` will probably always return false.
2. It's simpler to have one function that sets an "open" position and another for the "closed" position. Toggling requires saving state on the current position.
3. The subsystem doesn't have a command that uses this.

Reply:
1f9c207b61b46cb288f70718849bb164f57c2a4b
c9723cebfaad5d317fab22d9d0a6d99f48f489dd

Reply:
And don't forget setting names of the commands.

Reply:
7b9a422b712b2368f7b87172d5f9503acce4b913

Reply:
You don't need the whole `new SequentialCommandGroup(…)`, the factory function I linked is shorter.
Also, using time-based stop conditions might be needed for some cases, but with closed-loop controlled systems, why not check directly what you want to achieve – reaching the setpoint? The time you wrote might be too short (bad) or too long (wasteful), or just irrelevant for shorter/longer movements.

Reply:
Also, I think we usually use `runOnce()` to set setpoints, @vichik123 will know better.

Reply:
b99d9ddb3e150d918dafa831a892bd8a66ce7246

Reply:
The servo can return its current angle, too, why not check that? You could though combine a timeout, too, for a case where the servo doesn't reach its setpoint for some reason.

Reply:
83847868e91cb33be521906782191b0385d2f324

Reply:
Nice!